### PR TITLE
Support for defaults when destructuring parameters

### LINKED ISSUE #296: Support for defaults when destructuring parameters
The following code:

```js
({a = 'b'}) => {}
```

When run through esprima returns:

```json
{
    "type": "Program",
    "body": [
        {
            "type": "ExpressionStatement",
            "expression": {
                "type": "ArrowFunctionExpression",
                "id": null,
                "params": [
                    {
                        "type": "ObjectPattern",
                        "properties": [
                            {
                                "type": "Property",
                                "key": {
                                    "type": "Identifier",
                                    "name": "a"
                                },
                                "computed": false,
                                "value": {
                                    "type": "AssignmentPattern",
                                    "left": {
                                        "type": "Identifier",
                                        "name": "a"
                                    },
                                    "right": {
                                        "type": "Literal",
                                        "value": "b",
                                        "raw": "'b'"
                                    }
                                },
                                "kind": "init",
                                "method": false,
                                "shorthand": true
                            }
                        ]
                    }
                ],
                "defaults": [],
                "body": {
                    "type": "BlockStatement",
                    "body": []
                },
                "generator": false,
                "expression": false
            }
        }
    ],
    "sourceType": "script"
}
```

And when run back through escodegen, returns:

```js
({a}) => {};
```
i.e., the `value` of `a` is not recognized and the AssignmentExpression is lost

..

I was playing around with this, and I found that adding `expr.value` as a third parameter to `generatePropertyKey` and modifying like so makes it work

```diff
-        result.push(this.generateExpression(expr, Precedence.Sequence, E_TTT));
+        if (value.type === 'AssignmentPattern') {
+            result.push(this.AssignmentPattern(value, Precedence.Sequence, E_TTT));
+        }
+        else {
+            result.push(this.generateExpression(expr, Precedence.Sequence, E_TTT));
+        }
```

Thing is, I'm unsure if this could have other unintended consequences...the tests still pass, fwiw.

I can send along a PR with this change, and a simple object destructuring test...  thoughts?

### PR DESCRIPTION
**Pull request by tswaters**
Fixes #296

### CODE REVIEW

**Comment by benjamind** [comment]
Great work on this @tswaters!

Do you know how long this will be before we see a new 1.8.x release with this included? Could really use this.

**Comment by tswaters** [comment]
I'm not sure... I'm not a contributer to escodegen... just noticed a problem and created a pull request.

**Comment by michaelficarra** [comment]
Sorry, I've been meaning to review this for a while now. I'll get to it (and if all's good, make a release) ASAP.

**Comment by DavidGDD** [comment]
I'm using Istanbul (which uses escodegen) and i've got the same problem with destructuring... Will this pull-requests be merged soon?

File: escodegen.js [hunk 1]
```diff
@@ -980,14 +980,20 @@
...
 
         if (computed) {
             result.push('[');
         }
 
-        result.push(this.generateExpression(expr, Precedence.Sequence, E_TTT));
+        if (value.type === 'AssignmentPattern') {
+            result.push(this.AssignmentPattern(value, Precedence.Sequence, E_TTT));
+        }
+        else {
```
Comment (michaelficarra) [inline escodegen.js]: I don't like this style, but I'm not going to let it block merging this PR.

**Comment by michaelficarra** [comment]
LGTM. Sorry for taking this long to review.

**Comment by benjamind** [comment]
Now this is merged, can you publish a new 1.8.x release to npm?

**Comment by michaelficarra** [comment]
Published `1.8.1`. Thanks for the reminder.

**Comment by benjamind** [comment]
Nice work.  Thanks!

**Comment by DavidGDD** [comment]
Thanks a lot!

**Pull request merged**